500, adding entry point to inspect command

ErnestMatskevich opened the pull request:
Renamed branch to #500 for tracking issue
Added test for Inspect.java file

ErnestMatskevich committed:
feat: add  inspection server with akes framework

ErnestMatskevich committed:
Update pom.xml

ErnestMatskevich committed:
Add unit test and qulice

Added:
qulice plugin in pom.xml
test_inspect.js file with unit test

For testing installed in dev dependencies:
sinon
proxyquire

ErnestMatskevich committed:
add license in package-info.java

ErnestMatskevich committed:
Merge branch 'objectionary:master' into 500

ErnestMatskevich committed:
Add simple test for Inspect.java

add stronger signatire for main() in Inspect.java

add test for Inspect.java

ErnestMatskevich commented:
I added simple test for `Inspect.java` file, but when run qulice I faced with tha same warnings as in previous PR:

```
[WARNING] Unused declared dependencies found:
org.eolang:eo-runtime:jar:0.56.2:compile
org.junit.jupiter:junit-jupiter-api:jar:5.12.2:compile
org.junit.jupiter:junit-jupiter-engine:jar:5.12.2:compile
org.takes:takes:jar:1.24.6:compile
```

Also I would like to clarify how to build `inspect.jar` with Maven. Since currently there is single `pom.xml` file, Maven makes `eoc.jar` file as it writen in its <artifactId>. For complete my task I need to create my own file `inspect.jar`. I would like to use `<execution>` in `pom.xml` to separate creating `eoc.jar` and my `inspect.jar`. Is it correct idea? I think that after separate creating `inspect.jar` there will be no more warning about unused dependencies, because qulice will be placed in `<execution>` of inspect part and will not see junit and eo-runtime dependecies. @yegor256 What do you think about use _<execution>_ in `pom.xml`?

There is 1 falling check in [eslint workflow](https://github.com/objectionary/eoc/actions/runs/15198624288/job/42748385329?pr=528) Is it on my side? If it is how can I fix that?

yegor256 commented:
@ErnestMatskevich I fixed eslint, please merge master into your branch

ErnestMatskevich committed:
Merge branch 'master' into 500

ErnestMatskevich committed:
Add profile to creating inspect.jar file

Added profile section for creating independing inspect.jar file. Placed Takes and qulice inside profile

yegor256 commented:
@volodya-lombrozo can you please check this design?

volodya-lombrozo reviewed (changes requested):
@ErnestMatskevich 

Thank you for the contribution! I have a few important notes related to the PR itself. Highly recommend reading ["Helping others review your changes"](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/getting-started/helping-others-review-your-changes). Pay attention to  the [Write small pull requests](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/getting-started/helping-others-review-your-changes#write-small-pull-requests) and the [Provide context and guidance](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/getting-started/helping-others-review-your-changes#provide-context-and-guidance) sections. Could you also please make the title of this PR a bit more meaningful, rather than just "500", please?

volodya-lombrozo reviewed (changes requested):
@yegor256 

### `inspect` -> `debug`

As I understand it, you want to write a debugger. Why not call it `eoc debug` then? Why should we hide this information under the misleading `inspect` command?

## Separate Project

Why do we need to implement the entire debugger in the `eoc` repository? Maybe it's better to move it to a separate repo, as [other projects](https://github.com/go-delve/delve) do?
Here, we can just leave a thin `eoc debug` command just to invoke this debugger.
Moreover, `eoc` is written in JS.  Adding `Java` would create a some sort of mess. 

What do you think?

yegor256 commented:
@volodya-lombrozo how would you call a tool that allows you go into a working Java virtual machine, find any object and see its details? Would it be a debugger?

Indeed, this tool deserves a separate project/repository, but maybe a bit later, when its first version is ready. So far, it's an early draft.

ErnestMatskevich committed:
Merge branch 'master' into 500

ErnestMatskevich committed:
fix: add missing comma in package.json

ErnestMatskevich commented:
@volodya-lombrozo Thank you for your review feedback. I've made the following improvements:

1. Updated PR title to clearly state the functionality.
2. Will follow up with smaller, focused PRs containing detailed descriptions

For the current changes in [pom.xml](https://github.com/objectionary/eoc/blob/5e205ea31878c21b84df4216f7ceb936213b3d59/mvnw/pom.xml) added Maven profile section to build self-contained `inspect.jar`. 

inspect command can be run with help 
```bash 
node src/eoc.js inspect 
``` 
or it is possible just build `inspect.jar` with help Maven and use `-Pinspect` for indepening package/compile.

Could you please review this profile configuration approach.

yegor256 commented on src/commands/inspect.js:
```diff
@@ -0,0 +1,78 @@
+module.exports = function(opts) {
+  return new Promise((resolve, reject) => {
+    const mvnDir = path.join(process.cwd(), 'mvnw');
+    console.info('Building EO inspect server with Maven...');
```
@volodya-lombrozo @ErnestMatskevich indeed, it's better to build it once, when we package eoc. Then, simply run it.

ErnestMatskevich committed:
Merge branch 'objectionary:master' into 500

ErnestMatskevich committed:
Added asserts to test

Added asserts in InspectSmokeTest.java and changed Inspect.java to handle interrupt

ErnestMatskevich committed:
Update pom.xml

Delete qulice plugin and redundant resources section in pom.xml

ErnestMatskevich committed:
Rename variables in inspect.js

Simplified variable names according to review feedback

ErnestMatskevich committed:
fix: use object shorthand in test_inspect.js

Changed { spawn: spawn } to { spawn } to comply with ESLint rules

ErnestMatskevich commented:
@volodya-lombrozo I implemented 4 out of 5 feedback items from your review. Below is the current progress:

**Completed:**

- Added asserts in `InspectSmokeTest.java` at [this commit](https://github.com/objectionary/eoc/pull/528/commits/e9382b508a7ceda29f2935d91df4f0557cfd827d)
Also changed `Inspect.java` to handle interrupt server

- Removed redundant `<resources>` in `pom.xml` and `qulice` plugin (will add in new PR) at [this commit](https://github.com/objectionary/eoc/pull/528/commits/057fee345d010edcfa718a347a66855e57163f0d)

- Simplified variable names in `test_inspect.js` at [this commit](https://github.com/objectionary/eoc/pull/528/commits/314b44d33089b34424d914ad8672567d85b10220)
Fixed object-shorthand for ESLint at [last commit](https://github.com/objectionary/eoc/pull/528/commits/59df1e8df75851397ec3e6894692f185df657ca2)

**In progress:**
- Change Maven configuration to build java server once

Could you please review the implemented changes?

ErnestMatskevich committed:
Rewrite InspectTest with proper waiting

Rename test to InspcetTest.java

Add in pom.xml hamcrest dependency

Add adaptive waiting

ErnestMatskevich committed:
Update Inspect.java

Delete unused import
Remove redundant variable body
Simplified FkRegex with lambda

ErnestMatskevich commented on mvnw/src/main/java/org/eolang/Inspect.java:
```diff
@@ -0,0 +1,55 @@
+    /**
+     * Private constructor.
+     */
+    private Inspect() {
```
@volodya-lombrozo I added `private` to constructor to follow `qulice` rules when qulice plugin was part of my code. Without private constructor, qulice spawned this error: 

All methods are static.  Consider using a utility class instead. Alternatively, you could add a private constructor or make the class abstract to silence this warning. (UseUtilityClass)

I agree that it is not utility class since there is a `main()`. I deleted misunderstinding comment and saved constructor itself to follow `qulice` rule in future, when plugin will be added in project.

All changes regarding to Inspect.java file were done in [this commit](https://github.com/objectionary/eoc/pull/528/commits/969012b59f9a7b9e95203347ec819be010db9895)

volodya-lombrozo replied:
@ErnestMatskevich Great thanks for the explanation. Now it's clear.

ErnestMatskevich committed:
Build inspect.jar once with Maven package

Add maven assembly plugin which build inspect.jar

Refactor inspect.js - no more build inspect.jar, just run it

Refacror test_inspect.jar - no more testing maven part since it is not job of inspect command now

ErnestMatskevich commented:
@yegor256 @volodya-lombrozo I changed `eoc inspect` command behaviour. Now command `mvn clean package` creates `inspect.jar` with server and `takes` dependency inside. `inspect.js` just load `inspect.jar` and does not build it every time. Done in [commit](https://github.com/objectionary/eoc/pull/528/commits/3e593de7135169dc3f88a07a6ff08e7e8843f613)

I wrote descriptor `src/assembly/inspect.xml` for `maven-assembly-plugin` which I use for build `inspect.jar` and excluded `.class` files with my code in `maven-jar-plugin` for `eoc.jar`.

I deleted Maven part from inspect.js since `inspect.jar` is building by `mvn clean package` and deleted checks for Maven in `test_inspect.js`


There are problems in workflow with build and itests:
- ### Build fails
1) [dataize at 0.56.2 version](https://github.com/objectionary/eoc/actions/runs/15377766908/job/43264609698?pr=528#step:7:222)
2) [dataize with command-line argument](https://github.com/objectionary/eoc/actions/runs/15377766908/job/43264609698?pr=528#step:7:235)
dataization for older version is passed, but this 2 steps fails. Is problem on my side? I did not edit default building except adding exclusions for files with my code to avoid include them in eoc.jar 

- ### itests fails
[Could't find object 'Φ.org.eolang.io'](https://github.com/objectionary/eoc/actions/runs/15377766911/job/43264609705?pr=528#step:6:29)

Could you please take a look for my [commit](https://github.com/objectionary/eoc/pull/528/commits/3e593de7135169dc3f88a07a6ff08e7e8843f613) and help to fix build

volodya-lombrozo commented on mvnw/src/test/java/org/eolang/InspectTest.java:
```diff
@@ -0,0 +1,66 @@
+            server.isAlive(),
+            is(true)
+        );
+        
```
@ErnestMatskevich Can we remove empty lines, please? It's a [code smell](https://www.yegor256.com/2014/11/03/empty-line-code-smell.html).

volodya-lombrozo commented on mvnw/src/test/java/org/eolang/InspectTest.java:
```diff
@@ -0,0 +1,66 @@
+        }
+        assertThat(
+            "Server thread should be running after start",
+            server.isAlive(),
```
@ErnestMatskevich Here you just check the the thread is alive. It tells nothing about the server availability.
https://docs.oracle.com/javase/8/docs/api/java/lang/Thread.html#isAlive--

ErnestMatskevich replied:
@volodya-lombrozo Yes, the purpose of this test now is just check that server is starting without any errors. The availability of server is cheking in `test_inspect.js` file.

volodya-lombrozo replied:
@ErnestMatskevich How do you know there aren't errors? It may be that the server didn't actually start, but the thread is still alive.

volodya-lombrozo commented on mvnw/src/test/java/org/eolang/InspectTest.java:
```diff
@@ -0,0 +1,66 @@
+                }
+            }
+        );
+        server.setDaemon(false);
```
@ErnestMatskevich Why do we need this? From https://stackoverflow.com/a/17878439/10423604:

> Any thread created by main thread, which runs main method in Java is by default non daemon because Thread inherits its daemon nature from the Thread which creates it i.e. parent Thread and since main thread is a non daemon thread, any other thread created from it will remain non-daemon until explicitly made daemon by calling setDaemon(true).

volodya-lombrozo reviewed (changes requested):
@ErnestMatskevich There are significant improvements. Thanks! Just a few small comments. 

As for filling build, I suggest creating a new issue for that.

ErnestMatskevich committed:
Clean up InspectTest.java

Remove empty lines as it code smell

Remove .setDaemon(false) as children thread is not daemon by default

ErnestMatskevich committed:
Merge branch 'master' into 500

ErnestMatskevich committed:
Update package-lock.json

ErnestMatskevich committed:
Update pom.xml

Set src/main/java directly to sourceDirectory to show Maven right path to Java code.

Instead add <source>${eo.generatedDir}</source> to org.codehaus.mojo for opportunity run the tests

ErnestMatskevich committed:
Update pom.xml

Add new execution to place EOfoo package to classes and inside eoc.jar

ErnestMatskevich committed:
Update pom.xml

Write cleaner exclude for ignoring only package-info.class for Java code of Inspector

ErnestMatskevich commented:
### Fixed all errors

1) Updated `package-lock.json` for includes `node-fetch` - [commit](https://github.com/objectionary/eoc/pull/528/commits/b6fca8b73cb3a0bd86b7e03a6c7dd4089194e755)

2) Set `src/main/java` in `<sourceDirectory>` of `<build`> section in `pom.xml` for help Maven to see Java code of Inspector and avoid errors with unrecognising Java code, when run other commands which set different directories in `<sourceDirectory>`. For takes all previous functionality of other part of `eoc` source code, in `build-helper-maven-plugin` from `pom.xml` was set original `<source>${eo.generatedDir}</source>`. - [commit](https://github.com/objectionary/eoc/pull/528/commits/ca58081f354608a1f9d8d533443e14e2e44da065)

3) The main problem was this line: `<exclude>**/package-info.class</exclude>` in `pom.xml`. 
The `**` part exclude all `package-info.class` in project, but should excludes specific` package-info.class` for Inspect part.
After refactoring line is: `<exclude>org/eolang/package-info.class</exclude>` and all checks are passed. - [commit](https://github.com/objectionary/eoc/pull/528/commits/5be56ca094c157cbf5bff36916b7e06337996992)

@yegor256 Please, take a look

ErnestMatskevich committed:
Merge branch 'master' into 500

ErnestMatskevich committed:
Remove empty lines in inspect.js

yegor256 commented:
@ErnestMatskevich we don't know for how long these files will stay with us. Once they are merged, they are with us (forever). It's better to place them correctly. The `mvnw/` is not a "Maven folder". It's "Maven wrapper". The files you commit have nothing to do with the wrapper, right?

ErnestMatskevich committed:
Create new top-level folder inspect

Extract source code files from mvnw and place them in inspect folder

ErnestMatskevich commented:
> @ErnestMatskevich we don't know for how long these files will stay with us. Once they are merged, they are with us (forever). It's better to place them correctly. The `mvnw/` is not a "Maven folder". It's "Maven wrapper". The files you commit have nothing to do with the wrapper, right?

Done in this [commit](https://github.com/objectionary/eoc/pull/528/commits/956fa4f5a052136004d01d716c39ccd1c1857462)

Yes, I do not work with wrapper directly, i just build `JAR` file for server

ErnestMatskevich committed:
Remove InspectTest.java

ErnestMatskevich committed:
Add puzzle in inspect.js

ErnestMatskevich committed:
Update inspect.js

Fix puzzle description

ErnestMatskevich commented:
@volodya-lombrozo Please, take a look. All tasks done, all checks passed

Favixx commented on package.json:
```diff
@@ -26,8 +26,12 @@
     "grunt-mocha-cli": "^7.0.0",
-    "mocha": "^11.5.0",
-    "patch-package": "^8.0.0"
+    "mocha": "10.8.2",
```
`mocha` is declared three times in `devDependencies` here (lines 29, 32, 34: `10.8.2`, `11.5.0`, `^11.5.0`). The last one silently wins, so the first two are dead — this looks like an unresolved merge conflict. Please keep a single `mocha` entry (`^11.5.0`) and drop the other two.

Favixx commented on src/commands/inspect.js:
```diff
@@ -0,0 +1,57 @@
+const path = require('path');
+const os = require('os');
+const readline = require('readline');
+const fetch = require('node-fetch');
```
`node-fetch` is `require`d here but it is **not** listed in `dependencies` in `package.json` — it only resolves today because it happens to be hoisted as a transitive dep (`^2.6.7` in the lockfile). This will break the moment the transitive tree changes or resolves to `node-fetch@3` (ESM-only, where `require()` throws). Please add `node-fetch` explicitly to `dependencies`, pinned to `^2`.

Favixx commented on src/commands/inspect.js:
```diff
@@ -0,0 +1,57 @@
+    ], { cwd: mvnDir, stdio: ['pipe', 'pipe', 'pipe'] });
+    server.stdout.setEncoding('utf8').on('data', d => console.log(`[SERVER] ${d}`));
+    server.stderr.setEncoding('utf8').on('data', d => console.error(`[SERVER ERROR] ${d}`));
+    setTimeout(() => {
```
Waiting a fixed `setTimeout(..., 2000)` for the server to come up is a race: on a slow/loaded machine the first request can fire before the port is bound (failure), and on a fast machine you always waste 2s. Better to poll the `/` endpoint until it answers (or read the server's stdout for a ready marker) and proceed on that signal.

Favixx commented on src/commands/inspect.js:
```diff
@@ -0,0 +1,57 @@
+  return new Promise((resolve, reject) => {
+    const mvnDir = path.join(process.cwd(), 'mvnw');
+    const jar = path.join(mvnDir, 'target', 'inspect.jar');
+    const server = spawn('java', [
```
No `server.on('error', ...)` handler. If `java` isn't on `PATH` or `inspect.jar` is missing, `spawn` emits an `error` event that is currently unhandled and will crash the process with an opaque stack. Since the surrounding Promise already exposes `reject`, wire the spawn error to `reject(err)` with a friendly message (e.g. "run `npx grunt` first to build inspect.jar").

Favixx commented on mvnw/pom.xml:
```diff
@@ -35,10 +35,15 @@
     </dependency>
+    <dependency>
+      <groupId>org.hamcrest</groupId>
+      <artifactId>hamcrest</artifactId>
```
The new `hamcrest` dependency has no `<scope>`, so it defaults to `compile` and ends up on the runtime classpath. Hamcrest is a test-only assertion library — please add `<scope>test</scope>`.

Favixx commented on mvnw/pom.xml:
```diff
@@ -35,10 +35,15 @@
   <build>
     <finalName>eoc</finalName>
-    <sourceDirectory>${eo.generatedDir}</sourceDirectory>
+    <sourceDirectory>../inspect/src/main/java</sourceDirectory>
```
Switching `<sourceDirectory>` from `${eo.generatedDir}` to `../inspect/src/main/java` repoints the **primary** compile root away from the generated EO sources, then re-adds the generated dir via build-helper `add-source`. This inverts the primary/secondary source roles and is fragile. Please confirm the generated EO sources still compile under every profile (not just the default `inspect` one), ideally with a comment explaining why the swap is needed.

Favixx commented on test/commands/test_inspect.js:
```diff
@@ -0,0 +1,44 @@
+    });
+    const inspectPromise = inspect({ stdin: input, stdout: out });
+    input.write('hello\n');
+    await new Promise(res => setTimeout(res, 5000));
```
Two hard `setTimeout(5000)` sleeps (10s total) make this test slow and flaky — it's implicitly coupled to the 2000ms startup delay inside `inspect.js`. Prefer event-driven waiting (resolve when the echoed response arrives) or `sinon` fake timers, so the test is deterministic and fast.

Favixx commented on test/commands/test_inspect.js:
```diff
@@ -0,0 +1,44 @@
+const stream = require('stream');
+const proxyquire = require('proxyquire');
+describe('inspect command (mocked)', () => {
+  it('should send text and receive echoed response, then exit cleanly', async () => {
```
Only the happy path is covered. There's no test for the `catch` branch (fetch failure → "Request failed"), for the spawn-error path, or for the Java server itself. At minimum a fetch-failure case would guard the error handling this command depends on.

Favixx reviewed (changes requested):
Thanks for the work on the `inspect` entry point. The overall shape (command → Java HTTP server via takes → assembly jar) is reasonable, but I don't think it's mergeable yet — there are a few correctness/build issues plus test-quality gaps. Details inline; most are quick to fix.

Main blockers:
1. `package.json` declares `mocha` three times with conflicting versions — looks like an unresolved merge.
2. `node-fetch` is `require`d but is not declared in `dependencies` (only present transitively).
3. Server startup relies on a fixed `setTimeout(2000)`, which is racy.
4. `hamcrest` is added at compile scope instead of test.

Happy to re-review once these are addressed.
